### ISSUE TITLE
Tacoton-2 plus World vocoder

### ISSUE THREAD

**Issue by begeekmyfriend** [Rayhane-mamah/Tacotron-2#304]
Hey I am glad to inform you that I have succeeded to merge Tacotron model with World vocoder and generated some evaluation results as follows. The results sound not bad but still not perfect. However it shows another way to train different feature parameters with Tacotron. The World vocoder is an open source project and thus everyone can use it for all. Moreover the quality of resynth results from that vocoder is better than that from Griffin-Lim since the three features (lf0[1], mgc[60] and ap[5]) contain not only magnitude spectrograms but also phase information. Furthermore the depth of the features is low enough that we do not need postnet for Tacotron model. The performance of training can be reduced to 0.7 second per step. The inference can also be quick enough even it only works on CPU. So it really worthes trying.

I would like to share my experimental source code with you as follows. Note that it currently only for Chinese mandarin. You may modify it for other languages:
[tacotron-world-vocoder branch](https://github.com/begeekmyfriend/tacotron/tree/mandarin-world-vocoder)
[Python-Wrapper-for-World-Vocoder](https://github.com/begeekmyfriend/Python-Wrapper-for-World-Vocoder)
[pysptk merlin-world-vocoder branch](https://github.com/begeekmyfriend/pysptk/tree/merlin-world-vocoder)
By the way you need use `python setup.py install` and the copy the so file manually into the system path for `pysptk` and python wrapper project.

Besides I also would like to provide two Python scripts for World vocoder resynth test.
[world_vocoder_resynth_scripts.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2982117/world_vocoder_resynth_scripts.zip)

@Rayhane-mamah Let us rock with it! And @r9y9 thanks for your `pysptk` project.
[world_vocoder_demo.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2713888/world_vocoder_demo.zip)
![image](https://user-images.githubusercontent.com/6031938/50501258-ca2d1a80-0a91-11e9-8937-d607bd96f52e.png)

**Comment by begeekmyfriend** [comment]
By the way you need use `python setup.py install` and the copy the so file into the system path for `pysptk` and python wrapper project.

**Comment by begeekmyfriend** [comment]
It gets better when more training now.
[world_vocoder_demo.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2714364/world_vocoder_demo.zip)

**Comment by m-toman** [comment]
This is great, also wanted to tackle this some time ago but was busy with other projects.

So you use mgc2sp and vice versa from SPTK as in the Merlin project and not the codec WORLD provides? (https://github.com/mmorise/World/blob/master/examples/codec_test/readandsynthesis.cpp)

I've tried the WORLD codec with Merlin and I found that the MGC parameterization performed better (also REAPER got rid of most of the V/UV errors) but I never dug deep into the reason for it.

**Comment by begeekmyfriend** [comment]
I am using early version of WORLD vocoder source from [merlin](https://github.com/CSTR-Edinburgh/merlin/tree/master/tools/WORLD) but not the latest version in mmorise's repo which seems difficult to pass the resynth test scripts provided by merlin. I have not still deep insight on it. But I have forked my own modifed early verion WORLD vocoder source on my [repo](https://github.com/begeekmyfriend/World/tree/merlin) which works for me.

**Comment by m-toman** [comment]
Thanks.
Perhaps worth to integrate the modifications into this repository... do you know where the criticial differences between the Keithitho repo and this one are?

**Comment by begeekmyfriend** [comment]
Here is the implementation on my T2 fork branch only for Chinese mandarin https://github.com/begeekmyfriend/Tacotron-2/tree/mandarin-world-vocoder

**Comment by shartoo** [comment]
Nice job,thanks for your sharing!

**Comment by QueenKeys** [comment]
> Here is the implementation on my T2 fork branch only for Chinese mandarin https://github.com/begeekmyfriend/Tacotron-2/tree/mandarin-world-vocoder

Hello, may I ask the last dimension of the bap feature extracted by pyworld is 513, then need to change the bap parameter num_bap = 5 in hparams to num_bap = 513?

**Comment by QueenKeys** [comment]
> Here is the implementation on my T2 fork branch only for Chinese mandarin https://github.com/begeekmyfriend/Tacotron-2/tree/mandarin-world-vocoder

Hello, may I ask the last dimension of the bap feature extracted by pyworld is 1025, then need to change the bap parameter num_bap = 5 in hparams to num_bap = 1025?

**Comment by begeekmyfriend** [comment]
@QueenKeys I am using WORLD vocoder from Merlin but not the latest version on the repo. So please type such command and you would get the right vocoder library.
```shell
git clone https://github.com/begeekmyfriend/Python-Wrapper-for-World-Vocoder.git
git submodule update --init
```

**Comment by QueenKeys** [comment]
> @QueenKeys I am using WORLD vocoder from Merlin but not the latest version on the repo. So please type such command and you would get the right vocoder library.
> 
> ```shell
> git clone https://github.com/begeekmyfriend/Python-Wrapper-for-World-Vocoder.git
> git submodule update --init
> ```
when type 'git submodule update --init', Is this normal?
子模组 'lib/World'（https://github.com/mmorise/World）未对路径 'lib/World' 注册
正克隆到 '/home/queen/document/Python-Wrapper-for-World-Vocoder/lib/World'...
子模组路径 'lib/World'：检出 'd7c03432d572c5a162edba9c611b3c8e367069a9'

**Comment by begeekmyfriend** [comment]
@QueenKeys You might use `world_vocoder_resynth_scripts.zip` provided on the 1st floor to testify if it has been installed successfully or not.

**Comment by begeekmyfriend** [comment]
@QueenKeys Did you checkout the right branch `mandarin-world-vocoder` for this test?

**Comment by begeekmyfriend** [comment]
I do not use world-v2.

**Comment by QueenKeys** [comment]
> I do not use world-v2.

Hello, you may have misunderstood what I mean. I didn't say that you are using world-v2, but you set num_bap to 5 in hparam.py, so I guess it is possible that you set bap as in world-v2, otherwise I am not really sure why you set num_bap to 5?

**Comment by OswaldoBornemann** [comment]
@begeekmyfriend It seems that in your `https://github.com/begeekmyfriend/Tacotron-2/tree/mandarin-world-vocoder`, the `hparams.py` does not have `cbhg` parameters. Does i do something wrong?

**Comment by begeekmyfriend** [comment]
Hi everyone, I have upgraded WORLD vocoder into the latest version where we can use havest instead of dio for F0 pitch extraction. The link adress is still on the 1st floor. Any suggestion is welcome!

**Comment by Edresson** [comment]
Hi, I've been training the model for Portuguese. I have the same problem reported above with the LJ Speech dataset, during eval (during training using the tacotron teacher) I get good results, but during the synthesis the results are bad. My base has 10 hours of audio and I trained approximately 272k steps, is it necessary to train more, or is there a problem with the model?

The results reported here were obtained during eval (during training using the tacotron teacher)?

**Comment by begeekmyfriend** [comment]
@Edresson You need to checkout the alignment like this https://github.com/mozilla/TTS/issues/9#issuecomment-473743232

**Comment by Edresson** [comment]
@begeekmyfriend  I upgraded the repository as described, but the network does not converge, I believe it can be overfit, I tried Tacatron with grinffin-lim and also did not have good results. The Tacotron does not seem to converge on my own dataset. With my own dataset I managed to get good results with DCTTS, but with the tacotron the results are very bad. Do you have any suggestion ?

**Comment by begeekmyfriend** [comment]
The `Griffin Lim` branch is only for Chinese mandarin. Did you change the dictionary of your own?
As for WORLD features in my tests. For some of dataset it can learn alignment quickly but for others it fails. I am still working around with it.

**Comment by Edresson** [comment]
@begeekmyfriend yes I changed but I trained the model few steps I believe that with Griffin Lim the tacotron with my dataset converge after many steps, using the DCTTS I needed 2000k steps to get good results. For Tacotron-World the loss varies a lot during the training, and does not learn the alimentation. I also tried using the DCTTS with World vocoder, and I have the same problem, the model does not converge. If you get new results please inform me, I'm working on it too, any progress I will report here.

**Comment by begeekmyfriend** [comment]
If it fails under `griffin lim` branch it might well your dataset is not good enough for TTS

**Comment by Edresson** [comment]
@begeekmyfriend  I agree with you however using DCTTS get good results, as tacotron is more powerful I believe you need more data, I will train a model with griffin lim to check if that is the problem.

**Comment by begeekmyfriend** [comment]
Latest commit https://github.com/begeekmyfriend/Tacotron-2/commit/e40a7b73ac31d299d731439fbabe8921b231a739
Any feedback is welcome!
![step-13000-align](https://user-images.githubusercontent.com/6031938/55057163-57890680-50a3-11e9-83fa-19510c787374.png)

**Comment by superhg2012** [comment]
![step-20000-eval-align](https://user-images.githubusercontent.com/8817309/55123359-29f09b80-513d-11e9-9a1e-04607c749a3b.png)

Hi, @begeekmyfriend , I am runing Tacotron2 + pyworld using Biaobei(10000) tts corpus. Why my above alignment result is not continuious?

**Comment by begeekmyfriend** [comment]
I forgot to tell you that for differnt dataset, adjust the `hp.max_frame_num` and `hp.max_text_length` adopted by guided attention as the alignment slope for better convergence.

**Comment by superhg2012** [comment]
@begeekmyfriend I increased max_frame_num to 900 and keep max_text_length=300, should I increase max_text_length ？ what's the relation between these twp params?

**Comment by begeekmyfriend** [comment]
@superhg2012 NO, Biaobei dataset contains shorter clips and texts. You need to reduce these lengths to adapt the best ratio for N:T.

**Comment by superhg2012** [comment]
@begeekmyfriend I get it! thanks!!

**Comment by superhg2012** [comment]
Hi, @begeekmyfriend During my training T2, the eval stop_token loss is increasing while the train stop_token_loss is decreasing.  I found that in my training corpus, there is no puncuntations while some of the sentences for eval in hparams contains puncuntations. Is this root cause?

**Comment by Edresson** [comment]
Hi @begeekmyfriend , the alignment looks good however, when I run synthesize.py, I get an audio with only noise,  without speech. Did you get good results using synthesize.py?
See below the images of the alignments.

During training:
![During train 57k](https://i.imgur.com/HnJzEIi.png)
During the eval (in training)
![During eval 57k](https://i.imgur.com/a91SZnx.png)

**Comment by begeekmyfriend** [comment]
Because the stop token loss did not reduce to zero. My test is still undergoing as well.

**Comment by superhg2012** [comment]
> Because the stop token loss did not reduce to zero. My test is still undergoing as well.

switch dio for f0 estimation to harvest. Does it work for synthesize improvement?

**Comment by begeekmyfriend** [comment]
You may test it by resynth script.
[world_resynth.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3021525/world_resynth.zip)

**Comment by begeekmyfriend** [comment]
I found that MSE fits WORLD features more than MAE does because the value scales of lf0, mgc and bap differ. And the attention can be kept through the whole training. MAE works well for mel spectrograms because it contains only one kind of feature. See https://github.com/begeekmyfriend/Tacotron-2/commit/5863d5513ed34f94711a310d57722d0b1f990264

**Comment by superhg2012** [comment]
@begeekmyfriend can you share a better synthesized sample audio?

**Comment by begeekmyfriend** [comment]
The demo has been shown on the floors where there is spectrogram graph. Maybe I need to reduce the frame period to obtain better quality. However I was told that the fidelity of sample is not as good as that from G&L. The quality is better indeed.

**Comment by begeekmyfriend** [comment]
By the way when you want to hear complete synthesized samples, please wait **until the stop token loss reduced to zero.**

**Comment by begeekmyfriend** [comment]
As for alignment, remember adapting your `max_text_length` and `max_frame_num` to the best ratio of N:T which depends on your dataset.

**Comment by begeekmyfriend** [comment]
![image](https://user-images.githubusercontent.com/6031938/55381145-e048ec00-5554-11e9-9422-49938ad337e5.png)

**Comment by mrgloom** [comment]
@begeekmyfriend
Is pretrained model compatible with this repo https://github.com/begeekmyfriend/tacotron/tree/mandarin-world-vocoder is available for test?

**Comment by begeekmyfriend** [comment]
Imcompatible. In fact mantaining both of those tacotron project would exhaust me. So I am fixed on my T2 fork currently.

**Comment by superhg2012** [comment]
@begeekmyfriend  your learning cuve is better than mine, great!!

**Comment by osamasabri** [comment]
@begeekmyfriend  thanks for your implementation 
i am synthesizing Arabic speech i got good alignments i reached 50k steps but there is a problem with the speed of the synthesized audio when i reduce it to half it becomes normal but still not very clear do you have any idea why this is happening

**Comment by begeekmyfriend** [comment]
Here is biaobei mandarin demo from T2 + WORLD. The f0 feature value prediction is tough for this model.
[xmly_biaobei_world.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3062978/xmly_biaobei_world.zip)

**Comment by sujeendran** [comment]
> Well my fork from Ito's repo is just experimental project for my own tests and it is easy to modify thanks to its less code. I have ported some T2 code (e.x. location sensitive attention, stop tokens and dropout etc.) on my T1 fork to see what would happen. Generally speaking there is less different moduls between these two repos. You might regard my T1 fork as a simplified version of this T2 project.

@begeekmyfriend Can you please tell what are the modifications/steps required in the current version of your T1 repo to make it run with LJ Speech Dataset?
I am sorry I am asking you this now because most of the steps are mixed up in the previous comments and I thought it would be helpful for others to have it in one place too.
Also does this T1 version run with your updated WORLD vocoder repo?

**Comment by begeekmyfriend** [comment]
T1 modification is just a trivial version. I am focus on T2 currently.

**Comment by sujeendran** [comment]
@begeekmyfriend  Thanks for the response. In that case, do you mind giving a concise list of steps required for LJ Speech dataset?

**Comment by begeekmyfriend** [comment]
https://github.com/begeekmyfriend/Tacotron-2/tree/griffin-lim

**Comment by sujeendran** [comment]
> https://github.com/begeekmyfriend/Tacotron-2/tree/griffin-lim

Sorry, but the synthesizer is still using Griffin-Lim algorithm in this branch right? I was hoping for your World Vocoder implementation for LJ Speech. Changes needed for Hparams and preprocessing?

**Comment by begeekmyfriend** [comment]
Sorry, I forgot it. You might diff the `griffin-lim` and `mandarin-griffin-lim` branch and merge it into `mandarin-world-vocoder` branch. By the way in my test, the evaluation from WORLD plus Tacotron did not work as well as that from Griffin-Lim.

**Comment by sujeendran** [comment]
> Sorry, I forgot it. You might diff the `griffin-lim` and `mandarin-griffin-lim` branch and merge it into `mandarin-world-vocoder` branch. By the way in my test, the evaluation from WORLD plus Tacotron did not work as well as that from Griffin-Lim.

Thanks! I actually wanted to try World Vocoder as GL is really slow. I have managed to perform optimizations in the Tacotron part of the model and it is performing faster than before on CPU, but GL is still slow and reducing the number of iterations just kills the quality. I will try your implementation and see if i can get a good result.

**Comment by begeekmyfriend** [comment]
Give up this solution and turn to [WaveRNN](https://github.com/fatchord/WaveRNN). Feel free to reopen this issue.

**Comment by byuns9334** [comment]
@begeekmyfriend Hi, thank you for your great work. Have you tried Tacotron2 + WORLD vocoder for long sentence? When I test WORLD vocoder on very long sentence (about 200 characters), it gets some uncomfortable sound (something like break sound) in the sound. Have you experienced this problem? and what do you think of the solution for this?
[long.wav.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/4573139/long.wav.zip)

**Comment by begeekmyfriend** [comment]
You might split that long sentence into two sentences according to the pause point.

**Comment by JJun-Guo** [comment]
> 

您好，[Python-Wrapper-for-World-Vocoder](https://github.com/begeekmyfriend/Python-Wrapper-for-World-Vocoder) 这个链接已经失效了，可以重新发一下吗？是否已经提供了pretrain model呢？

**Comment by begeekmyfriend** [comment]
@JJ-Guo1996 I am not engaging in TTS any more. And I do not remember how I deleted that repo you mentioned. Maybe there are  substitutes for the WORLD vocoder listed in the [README](https://github.com/mmorise/World) file. You can refer to it.
